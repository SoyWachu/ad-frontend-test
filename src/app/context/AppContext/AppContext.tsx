"use client";
import { createContext, useContext } from "react";
import type { AppState, AppAction } from "@/app/reducers/appReducer";

export const AppContext = createContext<{
  state: AppState;
  dispatch: React.Dispatch<AppAction>;
  fetchGames: (
    page?: number,
    genre?: string,
    append?: boolean
  ) => Promise<void>;
} | null>(null);

export function useApp() {
  const context = useContext(AppContext);
  if (!context) throw new Error("useApp debe usarse dentro de AppProvider");
  return context;
}
