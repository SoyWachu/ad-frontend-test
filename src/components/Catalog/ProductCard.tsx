"use client";
import Image from "next/image";
import { useApp } from "@/app/context/AppContext/AppContext";

interface ProductCardProps {
  readonly id: string;
  readonly imageUrl: string;
  readonly title: string;
  readonly genre: string;
  readonly price: number;
  readonly isNew?: boolean;
  readonly description: string;
}

export default function ProductCard({
  id,
  imageUrl,
  title,
  genre,
  price,
  isNew = false,
  description,
}: ProductCardProps) {
  const { state, dispatch } = useApp();
  const isInCart = state.cart.some((item) => item.id === id);

  const handleClick = () => {
    if (isInCart) {
      dispatch({ type: "REMOVE_ITEM", payload: id });
    } else {
      dispatch({
        type: "ADD_ITEM",
        payload: {
          id,
          quantity: 1,
          title,
          price,
          imageUrl,
          genre,
          description,
        },
      });
    }
  };

  return (
    <div className="border rounded-lg overflow-hidden shadow-sm hover:shadow-md transition-shadow duration-200">
      <div className="relative">
        <Image
          src={imageUrl}
          alt={title}
          width={400}
          height={250}
          className="w-full h-60 object-fill py-5 px-5 rounded-tl-[36px] rounded-tr-[36px]"
        />
        {isNew && (
          <span className="absolute top-8 left-8 bg-white text-black text-xs px-2 py-1 rounded font-normal border border-gray-500">
            New
          </span>
        )}
      </div>
      <div className="px-6">
        <p className="text-gray-500 text-xs uppercase">{genre}</p>
        <div className="flex justify-between align-center min-h-16">
          <h2 className="font-bold text-lg mt-1">{title}</h2>
          <p className="font-semibold mt-1">${price}</p>
        </div>
        <button
          onClick={handleClick}
          className={`mb-5 w-full border rounded px-4 py-2 font-semibold transition ${
            isInCart
              ? "bg-red-500 text-white hover:bg-red-600 border-red-500"
              : "border-gray-400 hover:bg-gray-100"
          }`}
        >
          {isInCart ? "REMOVE FROM CART" : "ADD TO CART"}
        </button>
      </div>
    </div>
  );
}
