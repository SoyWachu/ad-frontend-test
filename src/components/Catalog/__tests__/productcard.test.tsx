import { render, screen, fireEvent } from "@testing-library/react";
import ProductCard from "../ProductCard";
import React from "react";
import { useApp } from "@/app/context/AppContext/AppContext";

jest.mock("next/image", () => ({
  __esModule: true,
  default: (props: any) => <img {...props} alt={props.alt || "mocked image"} />,
}));

jest.mock("@/app/context/AppContext/AppContext", () => ({
  __esModule: true,
  useApp: jest.fn(),
}));

describe("ProductCard", () => {
  const mockDispatch = jest.fn();
  const mockFetchGames = jest.fn();

  const baseProps = {
    id: "1",
    imageUrl: "/game.png",
    title: "Cool Game",
    genre: "Action",
    price: 100,
    description: "An awesome game",
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders product info correctly", () => {
    (useApp as jest.Mock).mockReturnValue({
      state: { cart: [] },
      dispatch: mockDispatch,
      fetchGames: mockFetchGames,
    });

    render(<ProductCard {...baseProps} />);

    expect(screen.getByText("Cool Game")).toBeInTheDocument();
    expect(screen.getByText("Action")).toBeInTheDocument();
    expect(screen.getByText("$100")).toBeInTheDocument();
    expect(screen.getByRole("img")).toHaveAttribute("src", "/game.png");
  });

  it("shows 'New' badge when isNew is true", () => {
    (useApp as jest.Mock).mockReturnValue({
      state: { cart: [] },
      dispatch: mockDispatch,
      fetchGames: mockFetchGames,
    });

    render(<ProductCard {...baseProps} isNew />);

    expect(screen.getByText("New")).toBeInTheDocument();
  });

  it("adds item to cart when not in cart", () => {
    (useApp as jest.Mock).mockReturnValue({
      state: { cart: [] },
      dispatch: mockDispatch,
      fetchGames: mockFetchGames,
    });

    render(<ProductCard {...baseProps} />);

    fireEvent.click(screen.getByRole("button", { name: /ADD TO CART/i }));

    expect(mockDispatch).toHaveBeenCalledWith({
      type: "ADD_ITEM",
      payload: expect.objectContaining({
        id: "1",
        title: "Cool Game",
        price: 100,
        quantity: 1,
        imageUrl: "/game.png",
        genre: "Action",
        description: "An awesome game",
      }),
    });
  });

  it("removes item from cart when already in cart", () => {
    (useApp as jest.Mock).mockReturnValue({
      state: {
        cart: [
          {
            id: "1",
            title: "Cool Game",
            price: 100,
            quantity: 1,
            imageUrl: "/game.png",
            genre: "Action",
            description: "An awesome game",
          },
        ],
      },
      dispatch: mockDispatch,
      fetchGames: mockFetchGames,
    });

    render(<ProductCard {...baseProps} />);

    fireEvent.click(screen.getByRole("button", { name: /REMOVE FROM CART/i }));

    expect(mockDispatch).toHaveBeenCalledWith({
      type: "REMOVE_ITEM",
      payload: "1",
    });
  });

  it("matches snapshot", () => {
    (useApp as jest.Mock).mockReturnValue({
      state: { cart: [] },
      dispatch: mockDispatch,
      fetchGames: mockFetchGames,
    });

    const { container } = render(<ProductCard {...baseProps} />);
    expect(container).toMatchSnapshot();
  });
});
