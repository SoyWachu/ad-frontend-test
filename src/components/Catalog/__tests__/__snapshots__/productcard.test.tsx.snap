// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ProductCard matches snapshot 1`] = `
<div>
  <div
    class="border rounded-lg overflow-hidden shadow-sm hover:shadow-md transition-shadow duration-200"
  >
    <div
      class="relative"
    >
      <img
        alt="Cool Game"
        class="w-full h-60 object-fill py-5 px-5 rounded-tl-[36px] rounded-tr-[36px]"
        height="250"
        src="/game.png"
        width="400"
      />
    </div>
    <div
      class="px-6"
    >
      <p
        class="text-gray-500 text-xs uppercase"
      >
        Action
      </p>
      <div
        class="flex justify-between align-center min-h-16"
      >
        <h2
          class="font-bold text-lg mt-1"
        >
          Cool Game
        </h2>
        <p
          class="font-semibold mt-1"
        >
          $
          100
        </p>
      </div>
      <button
        class="mb-5 w-full border rounded px-4 py-2 font-semibold transition border-gray-400 hover:bg-gray-100"
      >
        ADD TO CART
      </button>
    </div>
  </div>
</div>
`;
